### PR TITLE
Added support for rmw events, subscriptions/publisher info by topic, and enclaves

### PR DESCRIPTION
**Pull request by kuznetsovmoci**
These functions are required by foxy

### CODE REVIEW

File: rmw_opendds_shared_cpp/include/rmw_opendds_shared_cpp/event_converter.hpp [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RMW_OPENDDS_SHARED_CPP__EVENT_CONVERTER_HPP_
+#define RMW_OPENDDS_SHARED_CPP__EVENT_CONVERTER_HPP_
+
+//#include "ndds_include.hpp"
```
Comment (mitza-oci) [inline rmw_opendds_shared_cpp/include/rmw_opendds_shared_cpp/event_converter.hpp]: remove these lines

File: rmw_opendds_shared_cpp/src/event.cpp [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+    // call get status with the appropriate mask
+    // get_status should fill the event with the appropriate status information
+    ret_code = custom_event_info->get_status(status_kind, event_info);
+    */
+  } else {
+    RMW_SET_ERROR_MSG_WITH_FORMAT_STRING("event %d not supported", event_handle->event_type);
+  }
+
+  // if ret_code is not okay, return error and set taken to false.
```
Comment (mitza-oci) [inline rmw_opendds_shared_cpp/src/event.cpp:91]: Is the comment telling us anything additional to just reading the simple line of code below?

File: rmw_opendds_shared_cpp/src/event_converter.cpp [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+{
+  return mask_map.at(event_t);
+}
+
+bool is_event_supported(const rmw_event_type_t event_t)
+{
+  return mask_map.count(event_t) > 0;
+}
+
+//rmw_ret_t check_dds_ret_code(const DDS::ReturnCode_t dds_return_code)
```
Comment (mitza-oci) [inline rmw_opendds_shared_cpp/src/event_converter.cpp]: remove?

**Review by mitza-oci — CHANGES_REQUESTED** [review]

File: rmw_opendds_shared_cpp/src/event.cpp [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "rmw/error_handling.h"
+#include "rmw/impl/cpp/macros.hpp"
+
+//#include "rmw_connext_shared_cpp/connext_static_event_info.hpp"
```
Comment (mitza-oci) [inline rmw_opendds_shared_cpp/src/event.cpp]: remove

File: rmw_opendds_shared_cpp/src/event.cpp [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+  // TODO: uncomment and remove next line
+  //rmw_ret_t ret_code = RMW_RET_UNSUPPORTED;
+  rmw_ret_t ret_code = RMW_RET_OK;
+
+  // check if we support the input event type
+  if (is_event_supported(event_handle->event_type)) {
+    // lookup status mask from rmw_event_type
+    CORBA::ULong status_kind = get_status_kind_from_rmw(event_handle->event_type);
+
+    // TODO: implement
```
Comment (mitza-oci) [inline rmw_opendds_shared_cpp/src/event.cpp]: Is this something that will be implemented in this PR or at another time?

**Review by mitza-oci — CHANGES_REQUESTED** [review]

File: rmw_opendds_shared_cpp/include/rmw_opendds_shared_cpp/wait.hpp [hunk 6]
```diff
@@ -22,9 +22,41 @@
...
+  if (events) {
+    for (size_t i = 0; i < events->event_count; ++i) {
+      auto current_event = static_cast<rmw_event_t*>(events->events[i]);
+      RMW_CHECK_ARGUMENT_FOR_NULL(current_event->data, RMW_RET_INVALID_ARGUMENT);
+      //DDS::Entity* dds_entity = static_cast<ConnextCustomEventInfo*>(
+      //  current_event->data)->get_entity();
+      //if (!dds_entity) {
+      //  RMW_SET_ERROR_MSG("Event handle is null");
+      //  return RMW_RET_ERROR;
+      //}
```
Comment (adamsj-ros) [inline rmw_opendds_shared_cpp/include/rmw_opendds_shared_cpp/wait.hpp]: What's this commented section for? Is this a TODO or dead code we don't need anymore?

File: rmw_opendds_shared_cpp/include/rmw_opendds_shared_cpp/wait.hpp [hunk 7]
```diff
@@ -22,9 +22,41 @@
...
+
+      //DDS::StatusMask status_mask = dds_entity->get_status_changes();
+      bool is_active = false;
+
+      //if (is_event_supported(current_event->event_type)) {
+      //  is_active = static_cast<bool>(status_mask &
+      //    get_status_kind_from_rmw(current_event->event_type));
+      //}
+      // if status condition is not found in the active set
+      // reset the subscriber handle
```
Comment (adamsj-ros) [inline rmw_opendds_shared_cpp/include/rmw_opendds_shared_cpp/wait.hpp]: TODO or dead code?

File: rmw_opendds_shared_cpp/src/event.cpp [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+  RMW_CHECK_ARGUMENT_FOR_NULL(event_info, RMW_RET_INVALID_ARGUMENT);
+  RMW_CHECK_ARGUMENT_FOR_NULL(taken, RMW_RET_INVALID_ARGUMENT);
+
+  RMW_CHECK_TYPE_IDENTIFIERS_MATCH(
+    event handle,
+    event_handle->implementation_identifier,
+    implementation_identifier,
+    return RMW_RET_ERROR);
+
+  // TODO: uncomment and remove next line
```
Comment (adamsj-ros) [inline rmw_opendds_shared_cpp/src/event.cpp]: @kuznetsovmoci When do we uncomment?

**Review by mitza-oci — APPROVED** [review]

File: rmw_opendds_cpp/src/opendds_static_subscriber_info.cpp [hunk 9]
```diff
@@ -0,0 +1,97 @@
+// Copyright 2014-2019 Open Source Robotics Foundation, Inc.
```
Comment (adamsj-ros) [inline rmw_opendds_cpp/src/opendds_static_subscriber_info.cpp:1]: Do the OSRF copyrights belong in these repos?

File: rmw_opendds_shared_cpp/include/rmw_opendds_shared_cpp/guid_helper.hpp [hunk 10]
```diff
@@ -24,57 +24,24 @@ namespace DDS
...
-  return output << std::dec;
-}
-
-
-inline void DDS_InstanceHandle_to_GUID(DDS_GUID_t * guid, DDS_InstanceHandle_t instanceHandle)
-{
-  memcpy(guid->value, reinterpret_cast<DDS_Octet const *>(&instanceHandle), 16);
-}
-*/
 //  Taken from http://community.rti.com/comment/689#comment-689
```
Comment (adamsj-ros) [inline rmw_opendds_shared_cpp/include/rmw_opendds_shared_cpp/guid_helper.hpp:28]: keep?

File: rmw_opendds_shared_cpp/include/rmw_opendds_shared_cpp/opendds_static_event_info.hpp [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RMW_OPENDDS_SHARED_CPP__OPENDDS_STATIC_EVENT_INFO_HPP_
+#define RMW_OPENDDS_SHARED_CPP__OPENDDS_STATIC_EVENT_INFO_HPP_
+
+#include "rmw/ret_types.h"
+
+typedef struct OpenDDSCustomEventInfo
```
Comment (adamsj-ros) [inline rmw_opendds_shared_cpp/include/rmw_opendds_shared_cpp/opendds_static_event_info.hpp]: The name of this class makes it sound like it should be in the OpenDDS repo. Might be better for tracing code to insert "RMW" somewhere in the name.

File: rmw_opendds_shared_cpp/include/rmw_opendds_shared_cpp/qos.hpp [hunk 12]
```diff
@@ -132,9 +137,10 @@ rmw_qos_to_dds_qos(
     case RMW_QOS_POLICY_LIVELINESS_AUTOMATIC:
       dds_qos.liveliness.kind = DDS::AUTOMATIC_LIVELINESS_QOS;
       break;
-    case RMW_QOS_POLICY_LIVELINESS_MANUAL_BY_NODE:
-      dds_qos.liveliness.kind = DDS::MANUAL_BY_PARTICIPANT_LIVELINESS_QOS;
-      break;
+    // RMW_QOS_POLICY_LIVELINESS_MANUAL_BY_NODE has been depricated: use RMW_QOS_POLICY_LIVELINESS_MANUAL_BY_TOPIC below
+    //case RMW_QOS_POLICY_LIVELINESS_MANUAL_BY_NODE:
+    //  dds_qos.liveliness.kind = DDS::MANUAL_BY_PARTICIPANT_LIVELINESS_QOS;
+    //  break;
```
Comment (adamsj-ros) [inline rmw_opendds_shared_cpp/include/rmw_opendds_shared_cpp/qos.hpp]: TODO?

File: rmw_opendds_shared_cpp/include/rmw_opendds_shared_cpp/qos.hpp [hunk 13]
```diff
@@ -202,7 +208,9 @@ dds_qos_to_rmw_qos(
     rmw_qos.liveliness = RMW_QOS_POLICY_LIVELINESS_AUTOMATIC;
     break;
   case DDS::MANUAL_BY_PARTICIPANT_LIVELINESS_QOS:
-    rmw_qos.liveliness = RMW_QOS_POLICY_LIVELINESS_MANUAL_BY_NODE;
+    // RMW_QOS_POLICY_LIVELINESS_MANUAL_BY_NODE has been depricated: use RMW_QOS_POLICY_LIVELINESS_MANUAL_BY_TOPIC if manually asserted liveliness is needed.
```
Comment (adamsj-ros) [inline rmw_opendds_shared_cpp/include/rmw_opendds_shared_cpp/qos.hpp]:
```suggestion
    // RMW_QOS_POLICY_LIVELINESS_MANUAL_BY_NODE has been deprecated: use RMW_QOS_POLICY_LIVELINESS_MANUAL_BY_TOPIC if manually asserted liveliness is needed.
```

File: rmw_opendds_shared_cpp/include/rmw_opendds_shared_cpp/topic_cache.hpp [hunk 14]
```diff
@@ -127,49 +106,51 @@ class TopicCache
...
         "unique topic attempted to be added twice, ignoring");
       return false;
     }
-    topic_guid_to_info_[topic_guid] =
-      TopicInfo {participant_guid, topic_guid, topic_name, type_name};
-    participant_to_topic_guids_[participant_guid].insert(topic_guid);
+    endpoint_guid_to_info_[endpoint_guid] =
+      // Implement qos logic
+      //TopicInfo{ topic_name, type_name, participant_guid, endpoint_guid, qos_profile };
+      TopicInfo{ topic_name, type_name, participant_guid, endpoint_guid};
```
Comment (adamsj-ros) [inline rmw_opendds_shared_cpp/include/rmw_opendds_shared_cpp/topic_cache.hpp:119]: Looks like the commented line needs to be removed?

File: rmw_opendds_shared_cpp/include/rmw_opendds_shared_cpp/topic_endpoint_info.hpp [hunk 15]
```diff
@@ -0,0 +1,41 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (adamsj-ros) [inline rmw_opendds_shared_cpp/include/rmw_opendds_shared_cpp/topic_endpoint_info.hpp]: ?

File: rmw_opendds_shared_cpp/include/rmw_opendds_shared_cpp/topic_endpoint_info.hpp [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RMW_OPENDDS_SHARED_CPP__TOPIC_ENDPOINT_INFO_HPP_
+#define RMW_OPENDDS_SHARED_CPP__TOPIC_ENDPOINT_INFO_HPP_
+
+#include "rmw/topic_endpoint_info_array.h"
+#include "rmw_opendds_shared_cpp/visibility_control.h"
```
Comment (adamsj-ros) [inline rmw_opendds_shared_cpp/include/rmw_opendds_shared_cpp/topic_endpoint_info.hpp]: generally prefer <> when not local includes.

File: rmw_opendds_shared_cpp/include/rmw_opendds_shared_cpp/guid_helper.hpp [hunk 17]
```diff
@@ -24,57 +24,24 @@ namespace DDS
...
-/*
-inline void DDS_BuiltinTopicKey_to_GUID(DDS_GUID_t * guid, DDS_BuiltinTopicKey_t builtinTopicKey)
+inline void DDS_BuiltinTopicKey_to_GUID(DDS::GUID_t* guid, DDS::BuiltinTopicKey_t builtinTopicKey)
```
Comment (mitza-oci) [inline rmw_opendds_shared_cpp/include/rmw_opendds_shared_cpp/guid_helper.hpp:29]: I don't know how this is used by the rest of the RMW, but this is not a general way of translating a builtInTopicKey to a GUID.

File: rmw_opendds_shared_cpp/include/rmw_opendds_shared_cpp/opendds_static_event_info.hpp (hunk 11 above)
Comment (mitza-oci) [inline rmw_opendds_shared_cpp/include/rmw_opendds_shared_cpp/opendds_static_event_info.hpp]: `typedef struct` is from C, no need for it here.

File: rmw_opendds_shared_cpp/include/rmw_opendds_shared_cpp/qos.hpp [hunk 18]
```diff
@@ -132,9 +137,10 @@ rmw_qos_to_dds_qos(
     case RMW_QOS_POLICY_LIVELINESS_AUTOMATIC:
       dds_qos.liveliness.kind = DDS::AUTOMATIC_LIVELINESS_QOS;
       break;
-    case RMW_QOS_POLICY_LIVELINESS_MANUAL_BY_NODE:
-      dds_qos.liveliness.kind = DDS::MANUAL_BY_PARTICIPANT_LIVELINESS_QOS;
-      break;
+    // RMW_QOS_POLICY_LIVELINESS_MANUAL_BY_NODE has been depricated: use RMW_QOS_POLICY_LIVELINESS_MANUAL_BY_TOPIC below
```
Comment (mitza-oci) [inline rmw_opendds_shared_cpp/include/rmw_opendds_shared_cpp/qos.hpp]: remove these?

File: rmw_opendds_shared_cpp/include/rmw_opendds_shared_cpp/types.hpp [hunk 19]
```diff
@@ -46,38 +48,47 @@ class CustomDataReaderListener
     graph_guard_condition_(graph_guard_condition)
   {}
 
-  virtual void add_information(
-    const DDS::InstanceHandle_t & participant_instance_handle,
-    const DDS::InstanceHandle_t & instance_handle,
+  virtual bool add_information(
+    const DDS::GUID_t& participant_guid,
+    const DDS::GUID_t& instance_handle,
```
Comment (mitza-oci) [inline rmw_opendds_shared_cpp/include/rmw_opendds_shared_cpp/types.hpp]: Why is the name `instance_handle` if it's not one?

**Review by mitza-oci — CHANGES_REQUESTED** [review]
the Builtintopickey/instancehandle/guid stuff needs to be cleaned up

File: rmw_opendds_shared_cpp/include/rmw_opendds_shared_cpp/guid_helper.hpp (hunk 17 above)
Comment (kuznetsovmoci) [reply rmw_opendds_shared_cpp/include/rmw_opendds_shared_cpp/guid_helper.hpp:29]: This is about the only information I was able to find:
https://community.rti.com/kb/accessing-guid-connext-dds-entities

Comment (mitza-oci) [reply rmw_opendds_shared_cpp/include/rmw_opendds_shared_cpp/guid_helper.hpp:29]: I haven't read it but I have to assume it doesn't apply to OpenDDS.  As I put in the overall comment - this needs some design work.  Why is BuiltinTopicKey needed?  Is it sufficient to just use InstanceHandle and GUID?

Comment (kuznetsovmoci) [reply rmw_opendds_shared_cpp/include/rmw_opendds_shared_cpp/guid_helper.hpp:29]: How can I get Participant's InstanceHandle from on_data_available()?...

Comment (kuznetsovmoci) [reply rmw_opendds_shared_cpp/include/rmw_opendds_shared_cpp/guid_helper.hpp:29]: https://github.com/kuznetsovmoci/rmw_opendds/blob/6e9ad6da971c862390b3393d15d68a0e63fad309/rmw_opendds_shared_cpp/src/types/custom_publisher_listener.cpp#L60

Comment (kuznetsovmoci) [reply rmw_opendds_shared_cpp/include/rmw_opendds_shared_cpp/guid_helper.hpp:29]: Only way I can think of is storing it in the listener object, when node creates it...

Comment (mitza-oci) [reply rmw_opendds_shared_cpp/include/rmw_opendds_shared_cpp/guid_helper.hpp:29]: Which participant does it need to know about, and what exactly does it need (GUID)?

File: rmw_opendds_shared_cpp/include/rmw_opendds_shared_cpp/opendds_static_event_info.hpp (hunk 11 above)
Comment (mitza-oci) [reply rmw_opendds_shared_cpp/include/rmw_opendds_shared_cpp/opendds_static_event_info.hpp]: Not sure I agree here, we tend to use namespaces for that kind of thing.

File: rmw_opendds_shared_cpp/include/rmw_opendds_shared_cpp/guid_helper.hpp (hunk 17 above)
Comment (kuznetsovmoci) [reply rmw_opendds_shared_cpp/include/rmw_opendds_shared_cpp/guid_helper.hpp:29]: GUID of the DomainParticipant to which the DataReader (Writer) belongs

Comment (mitza-oci) [reply rmw_opendds_shared_cpp/include/rmw_opendds_shared_cpp/guid_helper.hpp:29]: I don't have enough context to understand what's going on here, and a PR review is not the right place to do design discussions.  Please start a separate document for this.

File: rmw_opendds_shared_cpp/include/rmw_opendds_shared_cpp/topic_cache.hpp (hunk 14 above)
Comment (kuznetsovmoci) [reply rmw_opendds_shared_cpp/include/rmw_opendds_shared_cpp/topic_cache.hpp:119]: Added 'TODO:' comment

File: rmw_opendds_shared_cpp/include/rmw_opendds_shared_cpp/opendds_static_event_info.hpp (hunk 11 above)
Comment (kuznetsovmoci) [reply rmw_opendds_shared_cpp/include/rmw_opendds_shared_cpp/opendds_static_event_info.hpp]: Agreed to keep the name, as OpenDDS prefix is used widely in rmw_opendds

File: rmw_opendds_shared_cpp/include/rmw_opendds_shared_cpp/guid_helper.hpp (hunk 10 above)
Comment (kuznetsovmoci) [reply rmw_opendds_shared_cpp/include/rmw_opendds_shared_cpp/guid_helper.hpp:28]: Will resolve when solution for DDS_BuiltinTopicKey_to_GUID is found

File: rmw_opendds_shared_cpp/include/rmw_opendds_shared_cpp/topic_endpoint_info.hpp (hunk 16 above)
Comment (kuznetsovmoci) [reply rmw_opendds_shared_cpp/include/rmw_opendds_shared_cpp/topic_endpoint_info.hpp]: Done

File: rmw_opendds_shared_cpp/include/rmw_opendds_shared_cpp/guid_helper.hpp (hunk 17 above)
Comment (kuznetsovmoci) [reply rmw_opendds_shared_cpp/include/rmw_opendds_shared_cpp/guid_helper.hpp:29]: Created Trello card:
https://trello.com/c/t3GAGc5a/85-find-alternative-to-ddsbuiltintopickeytoguid-function

Comment (jwillemsen) [reply rmw_opendds_shared_cpp/include/rmw_opendds_shared_cpp/guid_helper.hpp:29]: This is very likely something I noticed in the past when creating issue https://github.com/oci-labs/rmw_opendds/issues/2

Comment (kuznetsovmoci) [reply rmw_opendds_shared_cpp/include/rmw_opendds_shared_cpp/guid_helper.hpp:29]: Added this link to the Trello card

**Comment by adamsj-ros** [comment]
@kuznetsovmoci I believe some items have been postponed. Could you summarize those in as a comment in this PR?

@mitza-oci Have all your requested changes been addressed to your satisfaction?

**Comment by kuznetsovmoci** [comment]
There is one item remaining to be resolved after this PR is closed:
https://github.com/oci-labs/rmw_opendds/pull/44#pullrequestreview-436752098
We would need to find an alternative logic in DDS_BuiltinTopicKey_to_GUID() function, or use a key other than participant GUID in the map.
Design discussion is here: https://docs.google.com/document/d/1tg-AE0fldhpfoM-6Pf4d_h3_TnmAr4DAX4vAoKTC3zE/edit?ts=5eac9be7#bookmark=id.tktnw8o9ewtq

**Comment by adamsj-ros** [comment]
> There is one item remaining to be resolved after this PR is closed:
> [#44 (review)](https://github.com/oci-labs/rmw_opendds/pull/44#pullrequestreview-436752098)

Is that issue captured somewhere?

**Comment by kuznetsovmoci** [comment]
> > There is one item remaining to be resolved after this PR is closed:
> > [#44 (review)](https://github.com/oci-labs/rmw_opendds/pull/44#pullrequestreview-436752098)
> 
> Is that issue captured somewhere?

Yes, here: https://docs.google.com/document/d/1tg-AE0fldhpfoM-6Pf4d_h3_TnmAr4DAX4vAoKTC3zE/edit?ts=5eac9be7#bookmark=id.tktnw8o9ewtq. There is also a Trello card: https://trello.com/c/t3GAGc5a/85-find-alternative-to-ddsbuiltintopickeytoguid-function
